### PR TITLE
`Development`: Refactor quiz exercise participation page

### PR DESCRIPTION
**Pull request by rriyaldhi**
#### General

- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/guidelines/development-process/#naming-conventions-for-github-pull-requests).
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added multiple screenshots/screencasts of my UI changes.

### Motivation and Context

UI Components of Quiz Participation (both Course & Exam) will be reused for Quiz Exam (Quiz Questions from Quiz Pool that will be added in a separate PR). To avoid code duplication, the UI Components are decoupled from the QuizExercise object because Quiz Exam does not have the QuizExercise object.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Quiz Exercise (Can be any type: Synchronize, Batched, or Individual)

1. Log in to Artemis as a Student
5. Participate in the Quiz Exercise
6. Verify that the answers are saved by refreshing the page after submitting and ensure that the answers are saved
7. If Quiz is Batched or Individual, click **End Now** as Instructor. Or wait until the time ends for a Synchronized Quiz
8. Verify that the students can see the result
9. As an instructor, verify that the preview mode of the Quiz Exercise is still working by clicking the Preview Quiz button, submitting the answers, and verifying that the results are shown
10. As an instructor, verify that the solution mode of the Quiz Exercise is still working by clicking the Sample Solution button and verifying that the solutions are displayed

#### Exam Mode Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Exam with a Quiz Exercise (and additionally a Programming Exercise)

1. Log in to Artemis as a Student
2. Participate in the exam
3. Click the **Save** button
4. Refresh the page and verify that the answers are saved
5. Submit the exam
11. Log out and log in as an Instructor
12. Navigate to the Exam and click Assessment
13. Wait until the exam is finished and the grace period has ended, then click Evaluate Quiz Exercises
14. Log out and log in as a Student
15. Verify that the students can see the result

#### Performance Review
- [x] I confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2
#### Exam Mode Test
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| exam-participation.component.ts | 90.43% | ✅ |
| quiz-exam-submission.component.ts | 91.89% | ✅ |
| exam-participation-summary.component.ts | 93.06% | ✅ |
| quiz-exam-summary.component.ts | 92.45% | ✅ |
| quiz-participation.component.ts | 90.18% | ✅ |
| multiple-choice-question.component.ts | 93.75% | ✅ |
| quiz-scoring-info-student-modal.component.ts | 96.77% | ✅ |
| quiz.service.ts | 95% | ✅ |

### Screenshots

**Exam**
https://user-images.githubusercontent.com/10885503/234956543-c0235a61-8017-4c16-ae2d-b9bedbb347c2.mov

**Course**
https://user-images.githubusercontent.com/10885503/234956609-3c906fad-de56-4592-8351-50ce8b9bc97f.mov

### CODE REVIEW

**Review by undernagruzez — COMMENTED** [review]
Tested on ts3. For exercise mode, everything works as expected. While in exam mode, I opened 2 tabs with a quiz question. I ticked some answers on the first page and refreshed it, my answers stayed there. The second tab did not get the new state.

**Comment by rriyaldhi** [comment]
> Tested on ts3. For exercise mode, everything works as expected. While in exam mode, I opened 2 tabs with a quiz question. I ticked some answers on the first page and refreshed it, my answers stayed there. The second tab did not get the new state. Also, Artemis submitted blank results.

Hello @undernagruzez, I couldn't reproduce the issue. Here is my screencast:

https://github.com/ls1intum/Artemis/assets/10885503/c55d7f0c-ecf7-4320-a1c5-7ef3b461e703

Could you provide a screencast so that I know the exact sequence of actions that cause the issue?

**Comment by frederik-kt** [comment]
Tested on TS2. Quiz answers for all exercise types were not saved during the exam when reloading the page.

**Review by FelixTJDietrich — DISMISSED** [review]
Everything looks good to me and worked as expected on ts2 during the testing session. The quiz also saved correctly during exam mode as a student. Approved.

**Review by laadvo — DISMISSED** [review]
Tested on ts2 during the testing session, everything working as expected. Also tested the exam mode

**Review by greekgeek1 — DISMISSED** [review]
Tested on ts1. Worked as expected.

**Review by eceeeren — DISMISSED** [review]
Tested on TS3. Works as expected.

**Review by frederik-kt — APPROVED** [review]
Code LGTM.

**Review by YeetTheFirst21 — DISMISSED** [review]
Tested on TS2, all seems to be working as expected

**Review by Santia-go — APPROVED** [review]
Tested on TS1

Approved! 

I have tested verifying the following aspects: 

Quiz exercises: 
1. Creation and deletion of a quiz exercise work normally. 
2. Release for practice is also working 
3. Importing quizzes is also working
4. solving the quiz with batch, sync and individual mode is working. 
5. Previous button and sample solution buttons are working also. 
6. submit the preview as the instructor also works. 
7. The view of the results is also working. 
8. The scores are updated with the score of the quizzes. 
9. Students can see his score also on statistics and in "view result". 
10. The student can practice later and his store grade does not update.

In exam mode

1. Is possible to create, import and delete quiz exercises 
2. The exam is generated for students with the corresponding quiz for each student. 
3. The student can solve the quiz in the exam. 
4. The results of each question are saved (change the color of each question icon)
5. It is possible to submit the exam before.
6. The quiz appears in the summary of the exam at the end, and it can be exported as PDF normally in the summary. 
7. The grade of the exam is updated in the scores
8. The student can see his score

Good job here!

File: src/main/webapp/app/exam/participate/exam-participation.component.ts
Comment (MaximilianAnzinger) [inline src/main/webapp/app/exam/participate/exam-participation.component.ts:893]: Would it make sense to check if this is a valid cast?

Comment (rriyaldhi) [reply src/main/webapp/app/exam/participate/exam-participation.component.ts:893]: The check is expected to be done by the caller, which in this case is the *ngSwitchCase="QUIZ". This ensures that the exercise parameter is always of type QuizExercise.

**Review by DominikRemo — APPROVED** [review]
Tested on ts1 in testing session. Works as expected

File: src/main/webapp/app/exam/participate/summary/exercises/quiz-exam-summary/quiz-exam-summary.component.ts
Comment (simon-hng) [inline src/main/webapp/app/exam/participate/summary/exercises/quiz-exam-summary/quiz-exam-summary.component.ts:54]: Can be simplified
```suggestion
                this.studentParticipations.length > 0 && this.studentParticipations[0].results?.length 
```

File: src/main/webapp/app/exam/participate/summary/exercises/quiz-exam-summary/quiz-exam-summary.component.ts
Comment (simon-hng) [inline src/main/webapp/app/exam/participate/summary/exercises/quiz-exam-summary/quiz-exam-summary.component.ts:58]:
```suggestion
            this.result = this.submission.results?.length ? this.submission.results[0] : undefined;
```

**Review by simon-hng — COMMENTED** [review]
Left some optional suggestions and a questions.

Otherwise refactor looks good

**Review by frederik-kt — DISMISSED** [review]
Tested on TS1 during testing session. Results were saved correctly for a course and an exam exercise.

File: src/main/webapp/app/exam/participate/exam-participation.component.ts
Comment (MaximilianAnzinger) [reply src/main/webapp/app/exam/participate/exam-participation.component.ts:893]: Sorry, I should've been more specific. That's precisely what I wanted to suggest.

Comment (rriyaldhi) [reply src/main/webapp/app/exam/participate/exam-participation.component.ts]: Type guard is added before casting

File: src/main/webapp/app/exam/participate/exercises/quiz/quiz-exam-submission.component.ts
Comment (rriyaldhi) [reply src/main/webapp/app/exam/participate/exercises/quiz/quiz-exam-submission.component.ts]: The user of getExercise() method uses only the id and type.

**Review by MaximilianAnzinger — APPROVED** [review]
Tested on TS2 during testing session. Works as described 👍

**Review by pellumbbaboci — DISMISSED** [review]
Tested during testing session on ts2, everything works as expected.

**Review by tobias-lippert — APPROVED** [review]
Tested in testing session. Participating in quiz exercises worked flawlessly.

**Review by YeetTheFirst21 — DISMISSED** [review]
Retested on TS2, all seems to be working as expected.

**Review by Santia-go — DISMISSED** [review]
Re- approved after checking the last changes. 
All looks correct!

File: src/main/webapp/app/exam/participate/exercises/quiz/quiz-exam-submission.component.ts
Comment (krusche) [inline src/main/webapp/app/exam/participate/exercises/quiz/quiz-exam-submission.component.ts:50]: I don't like to have so many `undefined` inputs here. Does the page really work if those values are all undefined?

File: src/main/webapp/app/exam/participate/exercises/quiz/quiz-exam-submission.component.ts
Comment (krusche) [inline src/main/webapp/app/exam/participate/exercises/quiz/quiz-exam-submission.component.ts:56]: it's a bit cumbersome to define to pass so many attributes of the quiz. I understand that you need to provide an exercise independent way of passing values. However, can you not simply define an interface or super class which includes all those values so that you only need to pass one input?

File: src/main/webapp/app/exam/participate/exam-participation.component.ts
Comment (krusche) [inline src/main/webapp/app/exam/participate/exam-participation.component.ts:893]: this does not really look like good code to be honest

**Review by krusche — CHANGES_REQUESTED** [review]
I don't like the code changes, please reconsider alternative solutions

File: src/main/webapp/app/exam/participate/exam-participation.component.ts
Comment (rriyaldhi) [reply src/main/webapp/app/exam/participate/exam-participation.component.ts]: This is removed since the input was changed to the QuizConfiguration interface.

**Review by tobias-lippert — APPROVED** [review]
Tested in testing session on ts1. Quiz exercise participation worked flawlessly.

**Review by frederik-kt — DISMISSED** [review]
Tested on TS1, after refreshing the page for course and exam exercises, the results were still there.

**Review by MaximilianAnzinger — APPROVED** [review]
Tested on TS1 during testing session. Works as intended.

File: src/main/webapp/app/exam/participate/exercises/quiz/quiz-exam-submission.component.ts
Comment (krusche) [inline src/main/webapp/app/exam/participate/exercises/quiz/quiz-exam-submission.component.ts:74]: Are you sure this works? If the receiver only needs the id and nothing else from the exercise, we should consider changing the getter to `getExerciseId`, otherwise new consumers might think that certain values of the exercise are available while they are not

**Review by krusche — CHANGES_REQUESTED** [review]

File: src/main/webapp/app/exam/participate/exercises/quiz/quiz-exam-submission.component.ts
Comment (rriyaldhi) [reply src/main/webapp/app/exam/participate/exercises/quiz/quiz-exam-submission.component.ts:74]: The receiver only needs the exercise id and exercise type. So, yes this works.

Comment (krusche) [reply src/main/webapp/app/exam/participate/exercises/quiz/quiz-exam-submission.component.ts:74]: Yea this probably makes sense

Comment (rriyaldhi) [reply src/main/webapp/app/exam/participate/exercises/quiz/quiz-exam-submission.component.ts]: Changed the getExercise() method to getExerciseId() and getExerciseType().

**Comment by undernagruzez** [comment]
Retested on ts2. Works as expected.

**Review by TimOrtel — DISMISSED** [review]
Tested on ts2. The exam mode worked as expected.

**Review by lennart-keller — DISMISSED** [review]
Tested on ts2. Worked well

**Review by YeetTheFirst21 — DISMISSED** [review]
Retested on TS1, all seems to be working as expected

**Review by Santia-go — DISMISSED** [review]
Tested on TS1. 
- Everything is working as expected with quizzes in a course or exam 
- The creation and importing of quizzes and questions (MC, DnD, SA) are working
- The deletion of quizzes and questions is working. 
- Reevaluation is also working.
- The update of the grades for students in the statistics is also working. 
- The grade is updated in scores for the instructor's view.

**Review by YeetTheFirst21 — DISMISSED** [review]
Reapproved after conflict fix

**Review by MaximilianAnzinger — DISMISSED** [review]
Recent changes look good to me

**Review by greekgeek1 — APPROVED** [review]
Tested on ts1, reapprove

**Review by lennart-keller — APPROVED** [review]
Code lgtm :)

**Review by Santia-go — APPROVED** [review]
Re-approved after solving the merge conflict!

File: src/test/javascript/spec/component/exam/participate/exam-participation.component.spec.ts
Comment (MaximilianAnzinger) [inline src/test/javascript/spec/component/exam/participate/exam-participation.component.spec.ts:51]:
```suggestion
```
Unused Import

**Review by MaximilianAnzinger — CHANGES_REQUESTED** [review]

**Review by tobias-lippert — APPROVED** [review]
Reapprove; latest changes do not affect production code.

**Review by Santia-go — APPROVED** [review]
Reapproved! 

The changes for solving the merge conflict do not affect the behaviour previously tested.

**Review by MaximilianAnzinger — APPROVED** [review]
Code lgtm

**Review by lennart-keller — APPROVED** [review]
Reapprove. Code lgtm and changes made do not affect tested functionality.

**Review by YeetTheFirst21 — APPROVED** [review]
Reapproved after test line coverage

**Review by YeetTheFirst21 — DISMISSED** [review]
Approve again after migration fixes

**Review by pal03377 — APPROVED** [review]
Tested in testing session on TS2 without issues

**Review by tobias-lippert — APPROVED** [review]
Tested in testing session participating in a course quiz exercise and an exam quiz exercise worked flawlessly.

**Review by MaximilianAnzinger — DISMISSED** [review]
Tested on TS2
Code LGTM

**Review by rstief — APPROVED** [review]
Tested on TS2.

**Review by YeetTheFirst21 — DISMISSED** [review]
Reapproved after migration conflicts

**Review by Santia-go — DISMISSED** [review]
Tested on TS1 
Working as expected!

**Pull request merged**
